### DISCUSSION TITLE
Http 멱등성 ?

### DISCUSSION THREAD

**Discussion by jimi567** [2024-woowacourse-study/level-interview#118] [Network]
.

**Comment by J-I-H-O** [comment]
**<정의>**
HTTP Method 의 멱등성이란, **연산을 여러 번 적용 하더라도 결과가 항상 달라지지 않는 성질**을 뜻합니다. 즉, **여러 번 동일한 요청을 보냈을 때 서버에 미치는 영향이 동일한 경우 멱등**하다고 말할 수 있습니다.

**<HTTP Method에 따른 멱등성 보장 여부>**
`GET`, `HEAD`, `OPTIONS`, `PUT`, `DELETE` 메서드는 멱등성이 보장되고, `POST`, `PATCH` 메서드는 멱등성이 보장되지 않습니다.

- `GET` : 서버에 존재하는 리소스를 단순히 읽어오는 메서드이기 때문에 여러번 수행되어도 결과가 변하지 않음. `HEAD`, `OPTIONS` 또한 마찬가지
- `PUT` : 서버에 존재하는 리소스를 요청에 담긴 내용대로 통째로 대체하기 때문에, 여러번 수행되어도 같은 결과를 만들어냄
- `DELETE` : ‘존재하는 리소스를 삭제하는 경우’와 ‘존재하지 않는 데이터를 삭제하려고 시도하는 경우’ 모두 여러번 수행되더라도 같은 결과(리소스가 존재하지 않게 됨)를 만들어냄
- `POST` : 리소스를 생성하거나 변경할 때 사용되기 때문에 서버의 상태가 달라질 수 있음
- `PATCH` : 설계에 따라 멱등성을 보장할 수도, 보장하지 않을 수도 있음. 리소스의 일부에 대해 변화를 명령할 수 있기 때문에, 단일 호출과 다중 호출에 대한 서버의 상태가 달라질 수 있음 (예: PATCH 요청 시마다 특정 필드를 증가 시키도록 구현하거나, 값을 추가(append)하도록 구현할 수 있음)

**<주의점>**
멱등성을 따질 때에는 **사용자 요청에 의한 리소스의 상태만을 고려**합니다. 
자원이 존재하는 상태에서 DELETE 요청을 보낸 경우를 예로 들어보겠습니다. 처음에는 200 OK 상태 코드를 응답 받았지만, 동일한 요청을 반복하는 경우 404 NOT FOUND 상태 코드를 응답받을 수 있습니다.

이때 멱등성의 판별 기준은 **서버에 미치는 영향이 동일한지 여부**입니다. 요청에 따라 상태 코드를 다르게 응답한다 하더라도, 서버 입장에서는 리소스가 없는 상태를 유지하기 때문에 DELETE 메서드는 멱등합니다.

**<추가로 찾아볼만한 키워드>**
Safe Method, Cacheable Method

https://mangkyu.tistory.com/251
https://hudi.blog/http-method-idempotent/
https://www.rfc-editor.org/rfc/rfc7231#section-4.2.2

**Comment by le2sky** [comment]
연산을 여러 번 적용하더라도 결과가 달라지지 않는 성질을 멱등성이라고 합니다.
HTTP 메서드의 멱등성은 동일한 요청을 한번 보내는 것과 여러번 보내는 것이 서로 동일한 효과를 지니며,
서버의 상태도 동일하게 남을 경우에 멱등하다고 이야기할 수 있습니다.

HTTP 메서드의 멱등성이 필요한 이유는 요청 재시도 때문인데요.
멱등하다면 요청을 재시도할 때 같은 서버의 상태를 보장하기 때문에 문제가 없습니다.
반면, 멱등하지 않다면 재시도 요청시 중복 요청을 보내 문제를 발생 시킬 수 있습니다.

예를 들어, 사용자가 결제하는 시점에  타임아웃으로 인해 정상 응답을 못받는 상황을 생각해 볼 수 있습니다.
위 경우에서 멱등성하지 않은 결제 API 경우에는 결제가 성공했는지 수동으로 확인하고 재요청해야합니다.
반면, 멱등한 결제 API의 경우에는 안심하고 여러 번 요청할 수 있으며 중복 요청(따닥)으로 발생하는 문제(중복 결제)를 방지할 수 있습니다.

https://mangkyu.tistory.com/251
https://hudi.blog/http-method-idempotent/
https://docs.tosspayments.com/blog/what-is-idempotency
